engine/src/main/java/org/terasology/rendering/nui/layers/ingame/metrics/DebugMetricsSystem.java

```diff
@@ -0,0 +1,120 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.rendering.nui.layers.ingame.metrics;
+
+
+import org.terasology.entitySystem.systems.BaseComponentSystem;
+import org.terasology.entitySystem.systems.RegisterSystem;
+import org.terasology.registry.Share;
+import org.terasology.utilities.collection.CircularToggleSet;
+
+
+/**
+ * Manages an ordered set of MetricsMode instances.
+ *
+ *
+ * A number of default metrics modes is instantiated in the initialize() method and becomes immediately available by
+ * iterating through the set repeatedly invoking the toggle() method. The getCurrentMode() method can also be used to
+ * obtain the mode currently pointed at. Further modes can be added to the set via the register method, while registered
+ * modes can be removed via the unregister method. A convenience method unregisterAll() removes all registered modes at
+ * once. MetricsModes are currently used by the DebugOverlay instance, displaying runtime statistics on screen.
+ * unregister and unregisterAll functions.
+ */
+@RegisterSystem
+@Share(DebugMetricsSystem.class)
+public class DebugMetricsSystem extends BaseComponentSystem {
+
+    private final MetricsMode defaultMode = new NullMetricsMode();
+    private CircularToggleSet<MetricsMode> modes;
+    private MetricsMode currentMode;
+
+    @Override
+    public void initialise() {
+        modes = new CircularToggleSet<>();
+
+        register(defaultMode);
+        register(new RunningMeansMode());
+        register(new SpikesMode());
+        register(new AllocationsMode());
+        register(new RunningThreadsMode());
+        register(new WorldRendererMode());
+        register(new RenderingExecTimeMeansMode("Rendering - Execution Time: Running Means - Sorted Alphabetically"));
+        toggle();
+    }
+
+
+    /**
+     * Adds a MetricsMode instance to the set. Use the toggle() and getCurrentMode() methods to iterate over the set and
+     * obtain the MetricsMode instances.
+     * @param mode a MetricsMode instance
+     */
+    public boolean register(MetricsMode mode) {
+        return modes.add(mode);
+    }
+
+    /**
+     * Returns current mode, initializes a default {@link NullMetricsMode}, if currentMode is null
+     * @return current mode
+     */
+    public MetricsMode getCurrentMode() {
+        return currentMode;
+    }
+
+    /**
+     * Iterates through the set of registered MetricsMode and returns the first instance whose method isAvailable()
+     * returns true. Notice that this could be the MetricsMode that was current at the time this method was invoked.
+     * @returns a MetricsMode instance
+     */
+    public MetricsMode toggle() {
+        do {
+            currentMode = modes.toggle();
+        } while (!currentMode.isAvailable());
+
+        return currentMode;
```

Comment:
While testing DebugMetricsSystem, I noticed that WorldRendererMode in test mode, isAvailable() returns false. But by providing NullMetricsMode instance every time as a default even if it's cleared (it's always returns isAvailable() true), I believe it will not loop forever.